sklearn/metrics/_pairwise_fast.pyx

```diff
@@ -33,79 +29,3 @@ def _chi2_kernel_fast(floating[:, :] X,
                     if nom != 0:
                         res  += denom * denom / nom
                 result[i, j] = -res
-
-
-def _sparse_manhattan(
```

Comment:
I think this function is the simplest and the most frugal implementation for computing the pairwise manhattan distances on a pair of CSR datasets. Yet I think supporting all of the combinations (of metrics, {dense, sparse}², float32 and float64, etc.) by having such implementations is hardly maintainable.

The abstractions that we develop with Cython have a cost (that we can estimate against this implementation) but I think they ease future extensions.

What are people thinking of this? Are the Cython abstractions reasonable?

Reply:
As you said elsewhere, I think we need a few comparative benchmarks (dense and sparse) to answer that question.

I agree the tempita Cython-class oriented code offers more flexibility to support all combinations of memory representation / dtypes / metrics and would lean towards removing special cases if the performance impact is ok.